crates/polars-arrow/src/legacy/kernels/fixed_size_list.rs

```diff
@@ -1,73 +1,127 @@
-use polars_error::{polars_bail, PolarsResult};
-use polars_utils::index::NullCount;
-use polars_utils::IdxSize;
-
-use crate::array::{ArrayRef, FixedSizeListArray, PrimitiveArray};
-use crate::compute::take::take_unchecked;
-use crate::legacy::prelude::*;
-use crate::legacy::utils::CustomIterTools;
-
-fn sub_fixed_size_list_get_indexes_literal(width: usize, len: usize, index: i64) -> IdxArr {
-    (0..len)
-        .map(|i| {
-            if index >= width as i64 {
-                return None;
-            }
-
-            index
-                .negative_to_usize(width)
-                .map(|idx| (idx + i * width) as IdxSize)
-        })
-        .collect_trusted()
-}
-
-fn sub_fixed_size_list_get_indexes(width: usize, index: &PrimitiveArray<i64>) -> IdxArr {
-    index
-        .iter()
-        .enumerate()
-        .map(|(i, idx)| {
-            if let Some(idx) = idx {
-                if *idx >= width as i64 {
-                    return None;
-                }
+use polars_error::{polars_bail, PolarsError, PolarsResult};
 
-                idx.negative_to_usize(width)
-                    .map(|idx| (idx + i * width) as IdxSize)
-            } else {
-                None
-            }
-        })
-        .collect_trusted()
-}
+use crate::array::growable::make_growable;
+use crate::array::{Array, ArrayRef, FixedSizeListArray, PrimitiveArray};
+use crate::bitmap::BitmapBuilder;
+use crate::compute::utils::combine_validities_and;
+use crate::datatypes::ArrowDataType;
 
 pub fn sub_fixed_size_list_get_literal(
     arr: &FixedSizeListArray,
     index: i64,
     null_on_oob: bool,
 ) -> PolarsResult<ArrayRef> {
-    let take_by = sub_fixed_size_list_get_indexes_literal(arr.size(), arr.len(), index);
-    if !null_on_oob && take_by.null_count() > 0 {
-        polars_bail!(ComputeError: "get index is out of bounds");
+    if cfg!(debug_assertions) {
+        let msg = "fn sub_fixed_size_list_get_literal";
+        dbg!(msg);
+    }
+
+    let ArrowDataType::FixedSizeList(_, width) = arr.dtype() else {
+        unreachable!();
+    };
+
+    let width = *width;
+
+    let index = usize::try_from(index).unwrap();
+
+    if !null_on_oob && index >= width {
+        polars_bail!(
+            ComputeError:
+            "get index {} is out of bounds for array(width={})",
+            index,
+            width
+        );
     }
 
     let values = arr.values();
-    // SAFETY:
-    // the indices we generate are in bounds
-    unsafe { Ok(take_unchecked(&**values, &take_by)) }
```

Comment:
`list.get()` / `array.get()` were building selection indices and then calling gather with them - I've re-written them to use loops instead.